### PR TITLE
Fix path for Lynx parallel BMS system voltage

### LINKED ISSUE #1522: Lynx parallel BMS on WASM not displaying system voltage when BMS is pending
If one of the paralleled BMSes is 'pending' (waiting until its voltage level is in range with the system voltage), we show the system voltage in the menu of that BMS. However, on WASM, this field is not set properly:

image

On the local UI, the field is showing the value correctly.
See [original issue](https://github.com/victronenergy/gui-v2/issues/1269)

### PR DESCRIPTION
**Pull request by DanielMcInnes**
Fixes #1522

### CODE REVIEW

File: pages/settings/devicelist/battery/PageBattery.qml [hunk 1]
```diff
@@ -137,7 +137,9 @@ Page {
 
 				//% "System voltage"
 				text: qsTrId("devicelist_battery_system_voltage")
-				dataItem.uid: BackendConnection.serviceUidForType("battery.lynxparallel") + _n2kDeviceInstance.value + "/Dc/0/Voltage"
+				dataItem.uid: BackendConnection.type === BackendConnection.MqttSource
+							  ? "mqtt/battery/" + _n2kDeviceInstance.value  + "/Dc/0/Voltage"
+							  : BackendConnection.serviceUidForType("battery.lynxparallel") + _n2kDeviceInstance.value + "/Dc/0/Voltage"
```
Comment (DanielMcInnes) [inline pages/settings/devicelist/battery/PageBattery.qml:142]: @blammit this is one of those cases where the mqtt path is quite different to the dbus path.

dbus path:
dbus/com.victronenergy.battery.lynxparallel1/Dc/0/Voltage

mqtt path:
mqtt/battery/1/Dc/0/Voltage

I'm not sure how best to handle this. Who defines the mqtt paths? Maybe the real fix here is to change the mqtt path.

File: src/backendconnection.cpp [hunk 2]
```diff
@@ -581,8 +581,8 @@ QString BackendConnection::serviceUidForType(const QString &serviceType) const
 	//  - Mock: mock/com.victronenergy.system
 
 	return m_type == MqttSource
-			? QStringLiteral("mqtt/%1/0").arg(serviceType)
-			: QStringLiteral("%1/com.victronenergy.%2").arg(uidPrefix()).arg(serviceType);
+			? QStringLiteral("mqtt/%2/%1").arg(deviceInstance).arg(serviceType.split(".")[0])
```
Comment (DanielMcInnes) [inline src/backendconnection.cpp:584]: This fixes the uid for serviceTypes with the form "foo.bar". Currently, "battery.lynxparallel" is the only serviceType that calls this function with the form of "foo.bar", everywhere else that calls BackendConnection::serviceUidForType() does it with a simpler serviceType, that is not dot-separated.

The intent is for the behavior to remain unchanged for every other caller of BackendConnection::serviceUidForType()

**Comment by DanielMcInnes** [comment]
It occurs to me that dbus-flashmq (https://github.com/victronenergy/dbus-flashmq.git) is the canonical source for converting dbus paths to mqtt paths. I'm not sure it's worth changing to use this at this stage of the project, but it's useful to know anyway. See the following code snippet:
```
// shortservicename.cpp
std::string ShortServiceName::make_short(std::string service)
{
    if (service.find("com.victronenergy.") != std::string::npos)
    {
        const std::vector<std::string> parts = splitToVector(service, '.');
        service = parts.at(2);
    }
    else if (service.find(".") != std::string::npos || service.find("/") != std::string::npos)
    {
        throw ValueError("String doesn't look like com.victronenergy.something or something without dots or slashes.");
    }

    return service;
}
```

File: pages/settings/devicelist/battery/PageBattery.qml [hunk 3]
```diff
@@ -137,7 +137,7 @@ Page {
 
 				//% "System voltage"
 				text: qsTrId("devicelist_battery_system_voltage")
-				dataItem.uid: BackendConnection.serviceUidForType("battery.lynxparallel") + _n2kDeviceInstance.value + "/Dc/0/Voltage"
+				dataItem.uid: BackendConnection.serviceUidForType("battery.lynxparallel", _n2kDeviceInstance.value) + "/Dc/0/Voltage"
```
[the remark this answers is not in the document]
Comment (DanielMcInnes) [reply pages/settings/devicelist/battery/PageBattery.qml:140]: updated as suggested. Added unit tests for BackendConnection.serviceUidForType().

File: src/backendconnection.cpp [hunk 4]
```diff
@@ -658,5 +659,17 @@ QVariant BackendConnection::mockValue(const QString &uid) const
 	return QVariant();
 }
 
+BackendConnectionTester::BackendConnectionTester() {
```
Comment (blammit) [inline src/backendconnection.cpp:662]: `{` on the next line

File: pages/settings/devicelist/battery/PageBattery.qml [hunk 5]
```diff
@@ -138,16 +138,16 @@ Page {
...
 
 			ListTextItem {
-				id: numberOfBms
+				id: nrOfBmses
 				//% "Number of BMSes"
 				text: qsTrId("devicelist_battery_number_of_bmses")
 				dataItem.uid: root.battery.serviceUid + "/NumberOfBmses"
-				allowed: defaultAllowed && dataItem.isValid
+				allowed: defaultAllowed && root.isParallelBms
 			}
```
Comment (blammit) [inline pages/settings/devicelist/battery/PageBattery.qml:151]: The "Alarms" menu also needs a binding to `root.isParallelBms`.

Comment (DanielMcInnes) [reply pages/settings/devicelist/battery/PageBattery.qml:151]: done

File: src/backendconnection.cpp (hunk 4 above)
Comment (DanielMcInnes) [reply src/backendconnection.cpp]: Done

File: tests/backendConnection/tst_backendConnection.cpp [hunk 6]
```diff
@@ -0,0 +1,15 @@
+/*
```
Comment (blammit) [inline tests/backendConnection/tst_backendConnection.cpp:1]: Could you lower-case these file names to `tst_backendconnection.cpp/qml`?

File: pages/settings/devicelist/battery/PageBattery.qml (hunk 5 above)
Comment (blammit) [reply pages/settings/devicelist/battery/PageBattery.qml:151]: I don't see the change, was it left out?

Comment (DanielMcInnes) [reply pages/settings/devicelist/battery/PageBattery.qml:151]: are you talking about this thing in PageBattery.qml:
```			ListNavigationItem {
				text: CommonWords.alarms
				allowed: !root.isParallelBms
				onClicked: {
					Global.pageManager.pushPage("/pages/settings/devicelist/battery/PageBatteryAlarms.qml",
							{ "title": text, "bindPrefix": root.battery.serviceUid })
				}
			}
```
?

Comment (blammit) [reply pages/settings/devicelist/battery/PageBattery.qml:151]: Oh sorry, completely missed that.

**Review by blammit — APPROVED** [review]

**Pull request merged**